src/lib.rs

```diff
@@ -2,20 +2,61 @@
 #![allow(clippy::borrow_deref_ref)]
 
 use std::collections::HashSet;
-use std::thread;
 
-use fancy_regex::Regex;
+use pcre2::bytes::{Regex, RegexBuilder};
 use pyo3::exceptions;
 use pyo3::prelude::*;
 use pyo3::types::{PyBytes, PyList, PyTuple};
 use pyo3::PyResult;
 use rustc_hash::FxHashMap as HashMap;
 
-fn _byte_pair_merge(piece: &[u8], ranks: &HashMap<Vec<u8>, usize>) -> Vec<std::ops::Range<usize>> {
-    let mut parts: Vec<_> = (0..piece.len()).map(|i| i..i + 1).collect();
+fn _byte_pair_merge<T>(
+    piece: &[u8],
+    ranks: &HashMap<Vec<u8>, usize>,
+    f: impl Fn(std::ops::Range<usize>) -> T,
+) -> Vec<T> {
+    // This is a vector of (start, rank).
+    // The rank is of the byte pair starting at position start.
+    // The rank of the last item in the vector is not a valid value.
+    let mut parts: Vec<(usize, usize)> = (0..piece.len() + 1).map(|i| (i, usize::MAX)).collect();
+
+    // NOTE: using a macro here because a closure fails to get inlined
+    // according to optimization remarks.
+    macro_rules! get_rank {
```

Comment:
I think you can `#[inline(always)]` a closure, you just might need to surround it in braces

Reply:
Unfortunately, `error[E0658]: attributes on expressions are experimental`. So `error[E0518]: attribute should be applied to function or closure` is a little optimistic for now.

Reply:
Also, the best closure I can come up with is
```
let get_rank_skip = |parts: &Vec<(usize, usize)>, start_idx: usize, skip: usize|  {
    if (start_idx + skip + 2) < parts.len() {
        ranks
            .get(&piece[parts[start_idx].0..parts[start_idx + skip + 2].0])
            .map(|r| *r)
    } else {
        None
    }
};

let get_rank = |parts: &Vec<(usize, usize)>, start_idx: usize| {
    get_rank_skip(parts, start_idx, 0)
};
```
Passing `parts` by reference, even though it can be captured, seems to be necessary. Otherwise, the borrow checker will complain that we mutably borrow (when assigning to parts) while the closure has an immutable borrow via the captured reference. It's possible there is a better solution, but I am not aware of one.
I think the best option is to keep the macro.